compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala

```diff
@@ -83,188 +129,122 @@ class PCPCheckAndHeal(@constructorOnly ictx: Context) extends TreeMapWithStages(
+              tryHeal(tp.symbol, tp, pos)
+            case prefix: ThisType if !tp.symbol.isStatic && level > levelOf(prefix.cls) =>
+              tryHeal(tp.symbol, tp, pos)
+            case prefix: TermRef if tp.symbol.isSplice =>
+              // Heal explice type splice in the code
+              if level > 0 then getQuoteTypeTags.getTagRef(prefix) else tp
+            case prefix: TermRef if !prefix.symbol.isStatic && level > levelOf(prefix.symbol) =>
+              tryHeal(prefix.symbol, tp, pos)
             case _ =>
-              if (tp.symbol.is(Param)) tp
-              else mapOver(tp)
-          }
-        case tp: ThisType =>
-          assert(checkSymLevel(tp.cls, tp, pos).isEmpty)
-          mapOver(tp)
+              mapOver(tp)
+        case tp: ThisType if level != -1 && level != levelOf(tp.cls) =>
+          levelError(tp.cls, tp, pos, "")
         case tp: AnnotatedType =>
-          derivedAnnotatedType(tp, apply(tp.parent), tp.annot)
+          val newAnnotTree = transform(tp.annot.tree)
+          derivedAnnotatedType(tp, apply(tp.parent), tp.annot.derivedAnnotation(newAnnotTree))
         case _ =>
           mapOver(tp)
-      }
-    }
   }
 
-  /** Check reference to `sym` for phase consistency, where `tp` is the underlying type
-   *  by which we refer to `sym`. If it is an inconsistent type try construct a healed type for it.
-   *
-   *  @return `None` if the phase is correct or cannot be healed
-   *          `Some(tree)` with the `tree` of the healed type tree for `${implicitly[quoted.Type[T]]}`
-   */
-  private def checkSymLevel(sym: Symbol, tp: Type, pos: SourcePosition)(implicit ctx: Context): Option[Tree] = {
-    /** Is a reference to a class but not `this.type` */
-    def isClassRef = sym.isClass && !tp.isInstanceOf[ThisType]
-
-    /** Is this a static path or a type porjection with a static prefix */
-    def isStaticPathOK(tp1: Type): Boolean =
-      tp1.stripTypeVar match
-        case tp1: TypeRef => tp1.symbol.is(Package) || isStaticPathOK(tp1.prefix)
-        case tp1: TermRef =>
-          def isStaticTermPathOK(sym: Symbol): Boolean =
-            (sym.is(Module) && sym.isStatic) ||
-            (sym.isStableMember && isStaticTermPathOK(sym.owner))
-          isStaticTermPathOK(tp1.symbol)
-        case tp1: ThisType => tp1.cls.isStaticOwner
-        case tp1: AppliedType => isStaticPathOK(tp1.tycon)
-        case tp1: SkolemType => isStaticPathOK(tp1.info)
-        case _ => false
-
-    /* Is a reference to an `<init>` method on a class with a static path */
-    def isStaticNew(tp1: Type): Boolean = tp1 match
-      case tp1: TermRef => tp1.symbol.isConstructor && isStaticPathOK(tp1.prefix)
-      case _ => false
-
-    if (!sym.exists || levelOK(sym) || isStaticPathOK(tp) || isStaticNew(tp))
-      None
-    else if (!sym.isStaticOwner && !isClassRef)
+  /** Check phase consistency of terms and heal incosistent type references. */
```

Comment:
inconsistent

Also, maybe rename to `healTypeOfTerm`? i was confused in what way `healType` and `healTermType` are different.